component/module/string/string.go

```diff
@@ -65,24 +69,36 @@ func New(o component.Options, args Arguments) (*Component, error) {
 	flowRegistry := prometheus.NewRegistry()
 
 	c := &Component{
-		opts: o,
-		log:  o.Logger,
-
-		ctrl: flow.New(flow.Options{
-			ControllerID: o.ID,
-			LogSink:      logging.LoggerSink(o.Logger),
-			Tracer:       flowTracer,
-			Reg:          flowRegistry,
-
-			DataPath:       o.DataPath,
-			HTTPPathPrefix: o.HTTPPath,
-			HTTPListenAddr: o.HTTPListenAddr,
-
-			OnExportsChange: func(exports map[string]any) {
-				o.OnStateChange(Exports{Exports: exports})
-			},
-		}),
+		opts:    o,
+		log:     o.Logger,
+		exports: make(map[string]any),
 	}
+	c.ctrl = flow.New(flow.Options{
+		ControllerID: o.ID,
+		LogSink:      logging.LoggerSink(o.Logger),
+		Tracer:       flowTracer,
+		Reg:          flowRegistry,
+
+		DataPath:       o.DataPath,
+		HTTPPathPrefix: o.HTTPPath,
+		HTTPListenAddr: o.HTTPListenAddr,
+
+		OnExportsChange: func(exports map[string]any) {
```

Comment:
This won't work if an export gets removed, because the component will keep it in cache forever. 

The reason OnExportsChange provides a map because it's intended to be the declarative state of all the currently defined exports in the module, with the OnExportsChange callback being used as a thin layer to expose it back to the caller; I think fixing this at the component level is too high-level of a fix.